### PR TITLE
dpx_dpxconf, read_exact & mp_parse body

### PR DESCRIPTION
**Pull request by burrbull**
Depend on #148

### CODE REVIEW

File: dpx/src/dpx_dpxutil.rs [hunk 1]
```diff
@@ -60,21 +60,20 @@ pub struct ht_iter {
...
+        return (c - b'A' + 10) as i32;
     }
-    if c as i32 >= 'a' as i32 && c as i32 <= 'f' as i32 {
-        return c as i32 - 'a' as i32 + 10i32;
+    if c >= b'a' && c <= b'f' {
+        return (c - b'a' + 10) as i32;
     }
-    -1i32
+    -1
 }
```
Comment (chirsz-ever) [inline dpx/src/dpx_dpxutil.rs]: Maybe it would be better to use `i32::from_str_radix` of standard library.

File: dpx/src/dpx_dpxutil.rs [hunk 2]
```diff
@@ -87,8 +86,7 @@ pub unsafe extern "C" fn min4(mut x1: f64, mut x2: f64, mut x3: f64, mut x4: f64
     }
     v
 }
```
Comment (chirsz-ever) [inline dpx/src/dpx_dpxutil.rs]: `x1.min(x2).min(x3).min(x4)` is more simple, so as `max4`.

Comment (chirsz-ever) [reply dpx/src/dpx_dpxutil.rs]: Sorry, I pressed the review button incautiously

Comment (burrbull) [reply dpx/src/dpx_dpxutil.rs]: I've already removed `min4`/`max4` in #148 .

File: dpx/src/dpx_dpxutil.rs (hunk 1 above)
Comment (burrbull) [reply dpx/src/dpx_dpxutil.rs]: These function are not equivalent

**Comment by burrbull** [comment]
Rebased.
cc @cormacrelf

**Comment by burrbull** [comment]
@Mrmaxmeier 
Need yet one test.

File: dpx/src/dpx_mpost.rs [hunk 3]
```diff
@@ -3623,35 +1569,34 @@ unsafe fn do_operator(mut token: *const i8, mut x_user: f64, mut y_user: f64) ->
...
                 warn!("Unkown PostScript operator.");
-                dump(*start, next);
+                dump_slice(&start[..pos]);
                 error = 1i32
             } else if top_stack < 1024_u32 {
-                let fresh19 = top_stack;
-                top_stack = top_stack.wrapping_add(1);
-                stack[fresh19 as usize] = pdf_new_number(value);
-                *start = next
+                stack[top_stack as usize] = pdf_new_number(value);
```
Comment (cormacrelf) [inline dpx/src/dpx_mpost.rs:1597]: May as well redefine `top_stack` as `usize`? It is sure cast to usize a lot.

File: dpx/src/dpx_pdfobj.rs [hunk 4]
```diff
@@ -3095,38 +3096,31 @@ unsafe fn find_xref(handle: &mut InputHandleWrapper, mut file_size: i32) -> i32
  * This routine must be called with the file pointer located
  * at the start of the trailer.
  */
-unsafe fn parse_trailer(mut pf: *mut pdf_file) -> *mut pdf_obj {
+unsafe fn parse_trailer(mut pf: *mut pdf_file) -> Option<*mut pdf_obj> {
     /*
      * Fill work_buffer and hope trailer fits. This should
      * be made a bit more robust sometime.
      */
```
Comment (cormacrelf) [inline dpx/src/dpx_pdfobj.rs:3103]: The code here is still using the `static mut work_buffer_u8: [u8; 1024]`, which is unsafe as well as this concern about the trailer maybe not fitting. I don't see anything wrong with using a Vec here.

**Comment by Mrmaxmeier** [comment]
> @Mrmaxmeier
> Need yet one test.

187ed7b7 vs 906f15ae
```
reports/187ed7b7.jsonl reports/906f15ae.jsonl
summary:
  identical samples: 2446
  \>     successful: 1547
  different samples: 0
    missing samples: 0
```

File: dpx/src/dpx_pdfobj.rs (hunk 4 above)
Comment (burrbull) [reply dpx/src/dpx_pdfobj.rs:3103]: Let's do it in another PR.
And maybe you know. Is there easy way to grow `Vec` from `Read` with exact number of bytes?

Comment (cormacrelf) [reply dpx/src/dpx_pdfobj.rs:3103]: There’s `Read::read_to_end(&mut self, &mut Vec<u8>)`.

Comment (burrbull) [reply dpx/src/dpx_pdfobj.rs:3103]: `read_to_end` reads to end of file, not exact number of bytes.

Comment (burrbull) [reply dpx/src/dpx_pdfobj.rs:3103]: maybe `file.take(5).read_to_end(` is what I need

**Comment by burrbull** [comment]
@Mrmaxmeier Need last test.

**Comment by cormacrelf** [comment]
I know I reviewed the top_stack thing, but can you move that last commit to a new PR? This one is big enough and it passes so let’s cut it off there

Unless @Mrmaxmeier pulls through for us, in which case this is OK

**Comment by burrbull** [comment]
> I know I reviewed the top_stack thing, but can you move that last commit to a new PR? This one is big enough and it passes so let’s cut it off there

Ok.
Merge.

**Comment by burrbull** [comment]
See #169

**Pull request merged**